core/src/main/java/io/parsingdata/metal/data/ConcatenatedValueSource.java

```diff
@@ -16,63 +16,74 @@
+        if (length.compareTo(ZERO) <= 0) {
+            return complete(() -> output);
+        }
+        if (values.isEmpty()) {
+            throw new IllegalStateException("Data to read is not available (offset=" + offset + ";length=" + length + ";source=" + this + ").");
         }
-        final byte[] outputData = new byte[length.intValueExact()];
-        if (offset.add(length).compareTo(left.getLength()) <= 0) {
-            System.arraycopy(left.getValue(), offset.intValueExact(), outputData, 0, length.intValueExact());
-        } else if (offset.compareTo(left.getLength()) >= 0) {
-            System.arraycopy(right.getValue(), offset.subtract(left.getLength()).intValueExact(), outputData, 0, length.intValueExact());
-        } else {
-            final BigInteger leftPartLength = left.getLength().subtract(offset);
-            System.arraycopy(left.getValue(), offset.intValueExact(), outputData, 0, leftPartLength.intValueExact());
-            System.arraycopy(right.getValue(), 0, outputData, leftPartLength.intValueExact(), length.subtract(leftPartLength).intValueExact());
+        if (currentOffset.add(values.head.slice.length).compareTo(offset) <= 0) {
+            return intermediate(() -> getData(values.tail, currentOffset.add(values.head.slice.length), currentDest, offset, length, output));
         }
-        return outputData;
+        final BigInteger localOffset = offset.subtract(currentOffset).compareTo(ZERO) < 0 ? ZERO : offset.subtract(currentOffset);
+        final BigInteger toCopy = length.compareTo(values.head.slice.length.subtract(localOffset)) > 0 ? values.head.slice.length.subtract(localOffset) : length;
+        System.arraycopy(values.head.slice.getData(), localOffset.intValueExact(), output, currentDest.intValueExact(), toCopy.intValueExact());
+        return intermediate(() -> getData(values.tail, currentOffset.add(values.head.slice.length), currentDest.add(toCopy), offset, length.subtract(toCopy), output));
     }
 
     @Override
     protected boolean isAvailable(final BigInteger offset, final BigInteger length) {
-        return checkNotNegative(length, "length").add(checkNotNegative(offset, "offset")).compareTo(left.getLength().add(right.getLength())) <= 0;
+        return checkNotNegative(length, "length").add(checkNotNegative(offset, "offset")).compareTo(this.length) <= 0;
     }
 
     @Override
     public String toString() {
-        return getClass().getSimpleName() + "(" + left + "," + right + "))";
+        return getClass().getSimpleName() + "(" + values + "(" + length + "))";
     }
 
     @Override
     public boolean equals(final Object obj) {
         return Util.notNullAndSameClass(this, obj)
-            && Objects.equals(left, ((ConcatenatedValueSource)obj).left)
-            && Objects.equals(right, ((ConcatenatedValueSource)obj).right);
+            && Objects.equals(values, ((ConcatenatedValueSource)obj).values)
```

Comment:
Can I perhaps make a setup for a nice comparison utility :joy:

Reply:
Sounds like a great idea!